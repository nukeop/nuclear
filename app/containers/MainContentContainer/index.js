import React from 'react';
import { Route, Switch, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as Actions from '../../actions';

import MainLayout from '../../components/MainLayout';

import AlbumViewContainer from '../AlbumViewContainer';
import ArtistViewContainer from '../ArtistViewContainer';
import DashboardContainer from '../DashboardContainer';
import FavoritesContainer from '../FavoritesContainer';
import LyricsContainer from '../LyricsContainer';
import PlaylistsContainer from '../PlaylistsContainer';
import PlaylistViewContainer from '../PlaylistViewContainer';
import PluginsContainer from '../PluginsContainer';
import SearchResultsContainer from '../SearchResultsContainer';
import SettingsContainer from '../SettingsContainer';
import TagViewContainer from '../TagViewContainer';
import EqualizerViewContainer from '../EqualizerViewContainer';
import LibraryViewContainer from '../LibraryViewContainer';

import DownloadsContainer from '../../containers/DownloadsContainer';

class MainContentContainer extends React.Component {
  componentDidMount () {
    if (this.props.history &&
      this.props.location &&
      this.props.location.pathname === '/') {
      this.props.history.push('/dashboard');
    }
  }

  render () {
    return (
      <Route render={({ location }) => {
        return (
          <MainLayout>
            <Switch key={location.key} location={location}>
              <Route path='/album/:albumId' component={AlbumViewContainer} />
              <Route path='/artist/:artistId' component={ArtistViewContainer} />
              <Route path='/dashboard' component={DashboardContainer} />
              <Route path='/downloads' component={DownloadsContainer} />
              <Route path='/favorites/tracks' component={FavoritesContainer} />
              <Route path='/playlists' component={PlaylistsContainer} />
              <Route path='/playlist/:playlistId' component={PlaylistViewContainer} />
              <Route path='/plugins' component={PluginsContainer} />
              <Route path='/settings' component={SettingsContainer} />
              <Route path='/tag/:tagName' component={TagViewContainer} />
              <Route path='/search' component={SearchResultsContainer} />
              <Route path='/lyrics' component={LyricsContainer} />
              <Route path='/equalizer' component={EqualizerViewContainer} />
              <Route path='/library' component={LibraryViewContainer} />
            </Switch>
          </MainLayout>
        );
      }
      } />
    );
  }
}

function mapStateToProps () {
  return {};
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators(Actions, dispatch)
  };
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MainContentContainer));
