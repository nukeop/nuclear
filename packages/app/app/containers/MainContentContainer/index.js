import React from 'react';
import { Route, Switch, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as SearchActions from '../../actions/search';

import MainLayout from '../../components/MainLayout';

import AlbumViewContainer from '../AlbumViewContainer';
import ArtistViewContainer from '../ArtistViewContainer';
import DashboardContainer from '../DashboardContainer';
import DownloadsContainer from '../DownloadsContainer';
import EqualizerViewContainer from '../EqualizerViewContainer';
import FavoritesContainer from '../FavoritesContainer';
import LibraryViewContainer from '../LibraryViewContainer';
import LyricsContainer from '../LyricsContainer';
import PlaylistViewContainer from '../PlaylistViewContainer';
import PlaylistsContainer from '../PlaylistsContainer';
import PluginsContainer from '../PluginsContainer';
import SearchResultsContainer from '../SearchResultsContainer';
import SettingsContainer from '../SettingsContainer';
import TagViewContainer from '../TagViewContainer';
import VisualizerNode from '../VisualizerContainer/VisualizerNode';

class MainContentContainer extends React.Component {
  componentDidMount () {
    if (this.props.history &&
      this.props.location &&
      this.props.location.pathname === '/') {
      this.props.history.push('/dashboard');
    }
  }

  render () {
    return (
      <Route render={({ location }) => {
        return (
          <MainLayout>
            <Switch key={location.key} location={location}>
              <Route path='/album/:albumId' component={AlbumViewContainer} />
              <Route path='/artist/:artistId' component={ArtistViewContainer} />
              <Route path='/dashboard' component={DashboardContainer} />
              <Route path='/downloads' component={DownloadsContainer} />
              <Route path='/favorites/albums' component={FavoritesContainer} />
              <Route path='/favorites/tracks' component={FavoritesContainer} />
              <Route path='/favorites/artists' component={FavoritesContainer} />
              <Route path='/playlists' component={PlaylistsContainer} />
              <Route path='/playlist/:playlistId' component={PlaylistViewContainer} />
              <Route path='/plugins' component={PluginsContainer} />
              <Route path='/settings' component={SettingsContainer} />
              <Route path='/tag/:tagName' component={TagViewContainer} />
              <Route path='/search' component={SearchResultsContainer} />
              <Route path='/lyrics' component={LyricsContainer} />
              <Route path='/equalizer' component={EqualizerViewContainer} />
              <Route path='/visualizer' component={VisualizerNode} />
              <Route path='/library' component={LibraryViewContainer} />
            </Switch>
          </MainLayout>
        );
      }
      } />
    );
  }
}

function mapStateToProps () {
  return {};
}

function mapDispatchToProps (dispatch) {
  return {
    searchActions: bindActionCreators(SearchActions, dispatch)
  };
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MainContentContainer));
