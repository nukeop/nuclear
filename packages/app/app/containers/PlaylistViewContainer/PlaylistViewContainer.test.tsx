import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import { buildStoreState } from '../../../test/storeBuilders';
import { AnyProps, configureMockStore, setupI18Next, TestRouterProvider, TestStoreProvider } from '../../../test/testUtils';
import MainContentContainer from '../MainContentContainer';
import { onReorder } from '.';

jest.mock('fs');

describe('Playlist view container', () => {
  beforeAll(() => {
    setupI18Next();
  });

  it('should display a playlist', () => {
    const { component } = mountComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should play all tracks in the playlist', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByText('Play').click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist 1',
        name: 'test track'
      }),
      expect.objectContaining({
        artist: 'test artist 2',
        name: 'test track 22'
      })
    ]);
    expect(state.player.playbackStatus).toEqual('PLAYING');
  });

  it('should add all tracks in the playlist to the queue', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('more-button').click());
    await waitFor(() => component.getByText(/Add playlist to queue/i).click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist 1',
        name: 'test track'
      }),
      expect.objectContaining({
        artist: 'test artist 2',
        name: 'test track 22'
      })
    ]);
    expect(state.player.playbackStatus).toEqual('PAUSED');
  });

  it('should delete the playlist', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('more-button').click());
    await waitFor(() => component.getByText(/delete/i).click());

    const state = store.getState();
    expect(state.playlists.playlists).toEqual([]);
  });

  it('should export the playlist', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('more-button').click());
    await waitFor(() => component.getByText(/export/i).click());
    const state = store.getState();
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const remote = require('electron').remote;
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const fs = require('fs');
    const [playlist] = state.playlists.playlists;
    // check if the dialog was open
    expect(remote.dialog.showSaveDialog).toHaveBeenCalledWith({
      defaultPath: playlist.name,
      filters: [
        { name: 'file', extensions: ['json'] }
      ],
      properties: ['createDirectory', 'showOverwriteConfirmation']
    });
    expect(remote.dialog.showSaveDialog).toHaveBeenCalledTimes(1);
    // check if the playlist was properly exported 
    expect(fs.writeFile).toHaveBeenCalledWith(
      'downloaded_playlist',
      JSON.stringify(playlist, null, 2),
      expect.any(Function)
    );
  });

  it('should rename the playlist', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('rename-button').click());
    const input = component.getByPlaceholderText('Playlist name...');
    fireEvent.change(input, { target: { value: 'new name' } });
    await waitFor(() => component.getByText(/rename/i).click());

    const state = store.getState();
    expect(state.playlists.playlists[0].name).toEqual('new name');
  });

  it('should add a single track to the queue', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getAllByTestId('add-to-queue')[1].click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist 2',
        name: 'test track 22'
      })
    ]);
  });

  it('should play a single track', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getAllByTestId('play-now')[1].click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist 2',
        name: 'test track 22'
      })
    ]);
    expect(state.player.playbackStatus).toEqual('PLAYING');
  });

  it('should add a single track to favorites', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getAllByTestId('track-popup-trigger')[1].click());
    await waitFor(() => component.getByText(/add to favorites/i).click());

    const state = store.getState();
    expect(state.favorites.tracks).toEqual([
      expect.objectContaining({
        artist: 'test artist 2',
        name: 'test track 22'
      })
    ]);
  });

  it('should add a single track to downloads', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getAllByTestId('track-popup-trigger')[1].click());
    await waitFor(() => component.getByText(/download/i).click());

    const state = store.getState();
    expect(state.downloads).toEqual([
      expect.objectContaining({
        completion: 0,
        status: 'Waiting',
        track: expect.objectContaining({
          artist: 'test artist 2',
          name: 'test track 22'
        })
      })
    ]);
  });

  const mountComponent = (initialStore?: AnyProps) => {
    const initialState = initialStore ||
      buildStoreState()
        .withPlugins()
        .withPlaylists()
        .withConnectivity()
        .build();

    const history = createMemoryHistory({
      initialEntries: ['/playlist/0']
    });

    const store = configureMockStore(initialState);

    const component = render(
      <TestRouterProvider
        history={history}
      >
        <TestStoreProvider
          store={store}
        >
          <MainContentContainer />
        </TestStoreProvider>
      </TestRouterProvider >
    );

    return { component, history, store };
  };
});

describe('Playlist view container - utils', () => {
  let playlist = {
    tracks: [1, 2, 3, 4]
  };

  let newPlaylist = {};
  const updatePlaylist = playlistState => {
    newPlaylist = playlistState;
  };

  let reorder;
  
  beforeEach(() => {
    newPlaylist = {};
    playlist = {
      tracks: [1, 2, 3, 4]
    };
    reorder = onReorder(playlist, updatePlaylist);
  });
  it('should reorder tracks correctly (1)', () => {
    reorder(0, 2);
    expect(newPlaylist).toEqual({
      tracks: [2, 3, 1, 4]
    });
  });

  it('should reorder tracks correctly (2)', () => {
    reorder(1, 3);
    expect(newPlaylist).toEqual({
      tracks: [1, 3, 4, 2]
    });
  });

  it('should reorder tracks correctly (3)', () => {
    reorder(0, 1);
    expect(newPlaylist).toEqual({
      tracks: [2, 1, 3, 4]
    });
  });
});
