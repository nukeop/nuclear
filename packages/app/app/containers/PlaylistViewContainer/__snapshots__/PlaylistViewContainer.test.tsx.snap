// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist view container should display a playlist 1`] = `
<DocumentFragment>
  <div
    class="main_layout_container"
  >
    <div
      class="playlist_view_container"
    >
      <div
        class="playlist"
      >
        <div
          class="playlist_view_info"
        >
          <div>
            <img
              class="playlist_thumbnail"
              src="test thumbnail"
            />
          </div>
          <div
            class="playlist_header"
          >
            <label
              class="playlist_header_label"
            >
              Playlist
            </label>
            <div
              class="playlist_name"
            >
              test playlist
              <button
                aria-label="Rename this playlist"
                class="ui basic icon button nuclear button"
                data-testid="rename-button"
              >
                <i
                  aria-hidden="true"
                  class="pencil icon"
                />
              </button>
            </div>
            <div
              class="playlist_buttons"
            >
              <button
                class="ui pink circular button nuclear button play_button"
              >
                <i
                  aria-hidden="true"
                  class="play icon"
                />
                 Play
              </button>
              <button
                class="ui basic circular button nuclear button more_button"
                data-testid="more-button"
              >
                <i
                  aria-hidden="true"
                  class="ellipsis horizontal icon"
                />
              </button>
            </div>
          </div>
        </div>
        <table
          class="track_table"
          role="table"
        >
          <thead>
            <tr
              role="row"
            >
              <th
                colspan="1"
                role="columnheader"
              >
                 
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                <span
                  class="center_aligned"
                >
                  <i
                    aria-hidden="true"
                    class="hashtag icon"
                  />
                </span>
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                <span
                  class="center_aligned"
                >
                  <i
                    aria-hidden="true"
                    class="image icon"
                  />
                </span>
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                 
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Title
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Artist
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                Length
              </th>
              <th
                colspan="1"
                role="columnheader"
              >
                <span
                  class="select_header"
                >
                  <div
                    class="ui fitted checkbox nuclear checkbox"
                    style="cursor: pointer;"
                  >
                    <input
                      class="hidden"
                      readonly=""
                      tabindex="0"
                      title="Toggle All Rows Selected"
                      type="checkbox"
                      value=""
                    />
                    <label />
                  </div>
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            data-rbd-droppable-context-id="0"
            data-rbd-droppable-id="track_table"
            role="rowgroup"
          >
            <tr
              aria-describedby="rbd-hidden-text-0-hidden-text-0"
              class=""
              data-rbd-drag-handle-context-id="0"
              data-rbd-drag-handle-draggable-id="test track 0"
              data-rbd-draggable-context-id="0"
              data-rbd-draggable-id="test track 0"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                class="narrow delete_cell"
                role="cell"
              >
                <button
                  class="ui tiny basic circular icon button nuclear button borderless"
                >
                  <i
                    aria-hidden="true"
                    class="times icon"
                  />
                </button>
              </td>
              <td
                class="position_cell narrow"
                role="cell"
              >
                <button
                  class="ui pink tiny circular icon button nuclear button add_button"
                  data-testid="play-now"
                >
                  <i
                    aria-hidden="true"
                    class="play icon"
                  />
                </button>
                <span
                  class="position_cell_value"
                >
                  1
                </span>
              </td>
              <td
                class="thumbnail_cell narrow"
                role="cell"
              >
                <img
                  class="thumbnail"
                  src="test thumbnail"
                />
              </td>
              <td
                class="favorite_cell narrow"
                role="cell"
              >
                <button
                  class="ui tiny basic circular icon button nuclear button borderless"
                >
                  <i
                    aria-hidden="true"
                    class="heart outline icon"
                  />
                </button>
              </td>
              <td
                class="title_cell"
                role="cell"
              >
                <span
                  class="title_cell_content"
                >
                  <span
                    class="title_cell_value"
                  >
                    test track
                  </span>
                  <span
                    class="title_cell_buttons"
                  >
                    <button
                      class="ui tiny basic circular icon button nuclear button borderless"
                      data-testid="add-to-queue"
                    >
                      <i
                        aria-hidden="true"
                        class="plus icon"
                      />
                    </button>
                    <button
                      class="ui tiny basic circular icon button nuclear button borderless"
                      data-testid="track-popup-trigger"
                    >
                      <i
                        aria-hidden="true"
                        class="ellipsis horizontal icon"
                      />
                    </button>
                  </span>
                </span>
              </td>
              <td
                role="cell"
              >
                test artist 1
              </td>
              <td
                role="cell"
              />
              <td
                class="select_cell narrow"
                role="cell"
              >
                <div
                  class="ui fitted checkbox nuclear checkbox"
                  style="cursor: pointer;"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    title="Toggle Row Selected"
                    type="checkbox"
                    value=""
                  />
                  <label />
                </div>
              </td>
            </tr>
            <tr
              aria-describedby="rbd-hidden-text-0-hidden-text-0"
              class=""
              data-rbd-drag-handle-context-id="0"
              data-rbd-drag-handle-draggable-id="test track 22 1"
              data-rbd-draggable-context-id="0"
              data-rbd-draggable-id="test track 22 1"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                class="narrow delete_cell"
                role="cell"
              >
                <button
                  class="ui tiny basic circular icon button nuclear button borderless"
                >
                  <i
                    aria-hidden="true"
                    class="times icon"
                  />
                </button>
              </td>
              <td
                class="position_cell narrow"
                role="cell"
              >
                <button
                  class="ui pink tiny circular icon button nuclear button add_button"
                  data-testid="play-now"
                >
                  <i
                    aria-hidden="true"
                    class="play icon"
                  />
                </button>
                <span
                  class="position_cell_value"
                >
                  2
                </span>
              </td>
              <td
                class="thumbnail_cell narrow"
                role="cell"
              >
                <img
                  class="thumbnail"
                  src="test thumbnail 2"
                />
              </td>
              <td
                class="favorite_cell narrow"
                role="cell"
              >
                <button
                  class="ui tiny basic circular icon button nuclear button borderless"
                >
                  <i
                    aria-hidden="true"
                    class="heart outline icon"
                  />
                </button>
              </td>
              <td
                class="title_cell"
                role="cell"
              >
                <span
                  class="title_cell_content"
                >
                  <span
                    class="title_cell_value"
                  >
                    test track 22
                  </span>
                  <span
                    class="title_cell_buttons"
                  >
                    <button
                      class="ui tiny basic circular icon button nuclear button borderless"
                      data-testid="add-to-queue"
                    >
                      <i
                        aria-hidden="true"
                        class="plus icon"
                      />
                    </button>
                    <button
                      class="ui tiny basic circular icon button nuclear button borderless"
                      data-testid="track-popup-trigger"
                    >
                      <i
                        aria-hidden="true"
                        class="ellipsis horizontal icon"
                      />
                    </button>
                  </span>
                </span>
              </td>
              <td
                role="cell"
              >
                test artist 2
              </td>
              <td
                role="cell"
              />
              <td
                class="select_cell narrow"
                role="cell"
              >
                <div
                  class="ui fitted checkbox nuclear checkbox"
                  style="cursor: pointer;"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    title="Toggle Row Selected"
                    type="checkbox"
                    value=""
                  />
                  <label />
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
